slime/ray/rollout.py

```diff
@@ -459,6 +460,42 @@ def _log_rollout_data(rollout_id, args, samples, rollout_time):
     if args.rollout_num_gpus is not None:
         log_dict["perf/tokens_per_gpu_per_sec"] = sum(response_lengths) / rollout_time / args.rollout_num_gpus
     log_dict["perf/longest_sample_tokens_per_sec"] = max(response_lengths) / rollout_time
+    # For group based rollout, additionally log the raw reward statistics. 
+    raw_rewards = torch.tensor([sample.get_reward_value(args) for sample in samples], dtype=torch.float)
+    sample_indices = torch.tensor([sample.index for sample in samples])
+    if args.advantage_estimator in ["grpo", "gspo", "reinforce_plus_plus_baseline"]:
+        if len(raw_rewards) == args.n_samples_per_prompt * args.rollout_batch_size:
+            rewards = raw_rewards.reshape(-1, args.n_samples_per_prompt)
+            sample_indices = sample_indices.reshape(-1, args.n_samples_per_prompt)
+            # Make sure per group sample indices are identical. 
+            assert (sample_indices == sample_indices[:, 0].unsqueeze(1)).all(dim=1).all(), f" {sample_indices} Per group sample indices are not identical."
+            group_count, sample_per_group = rewards.shape
+            # Collect pass@k statistics for each group
+            all_zeros = rewards.eq(0).all(dim=-1)
+            all_ones = rewards.eq(1).all(dim=-1)
+            
+            # Calculate pass@k metrics dynamically for powers of 2
+            # pass@k is the probability that at least one of the first k samples passes
```

Comment:
for example, for a group size between 8 and 15, we log 
pass@1, pass@2, pass@4, pass@8